server/backends/invocationdb/invocationdb.go

```diff
@@ -44,36 +96,25 @@ func (d *InvocationDB) createInvocation(tx *db.DB, ctx context.Context, ti *tabl
 	if permissions == nil && d.env.GetConfigurator().GetAnonymousUsageEnabled() {
 		permissions = perms.AnonymousUserPermissions()
 	} else if permissions == nil {
-		return status.PermissionDeniedErrorf("Anonymous access disabled, permission denied.")
+		return false, status.PermissionDeniedErrorf("Anonymous access disabled, permission denied.")
 	}
 
 	ti.UserID = permissions.UserID
 	ti.GroupID = permissions.GroupID
 	ti.Perms = ti.Perms | permissions.Perms
-	return tx.Create(ti).Error
+	return d.registerInvocationAttempt(ctx, ti)
 }
 
-// InsertOrUpdateInvocation checks whether an invocation with the same primary
-// key as the given invocation exists. If no such invocation exists, it will
-// create a new row with the given invocation properties. Otherwise, it will
-// update the existing invocation. It returns whether a new row was created.
-func (d *InvocationDB) InsertOrUpdateInvocation(ctx context.Context, ti *tables.Invocation) (bool, error) {
-	created := false
-	err := d.h.TransactionWithOptions(ctx, db.Opts().WithQueryName("upsert_invocation"), func(tx *db.DB) error {
-		var existing tables.Invocation
-		if err := tx.Where("invocation_id = ?", ti.InvocationID).First(&existing).Error; err != nil {
-			if !db.IsRecordNotFound(err) {
-				return err
-			}
-			if err := d.createInvocation(tx, ctx, ti); err != nil {
-				return err
-			}
-			created = true
-			return nil
-		}
-		return tx.Model(&existing).Where("invocation_id = ?", ti.InvocationID).Updates(ti).Error
+// UpdateInvocation updates an existing invocation with the given
+// id and attempt number. It returns whether a row was updated.
+func (d *InvocationDB) UpdateInvocation(ctx context.Context, ti *tables.Invocation) (bool, error) {
+	updated := false
+	err := d.h.TransactionWithOptions(ctx, db.Opts().WithQueryName("update_invocation"), func(tx *db.DB) error {
+		result := tx.Where("`attempt` = ?", ti.Attempt).Updates(ti)
```

Comment:
nit: this is a little confusing because it relies on gorm magic. I think it would be clearer if it had invocation_id = ? and atttempt = ? so it was clear what was going on?